src/programQueueProcessor/eventProcessor.ts

```diff
@@ -60,15 +80,75 @@ export default ({
+
       await withRetry(async (retry, attemptIndex) => {
-        newResolvedIndex = await rollCallClient.createNewResolvableIndex(
-          programId.toLowerCase(),
-          true
-        );
-        logger.info(`obtained new index name: ${newResolvedIndex.indexName}`);
-        try {
-          await initIndexMapping(newResolvedIndex.indexName, esClient);
+        const existingIndexName = await getLatestIndexName(esClient, programId);
+        if (existingIndexName) {
+          const response = await getIndexSettings(esClient, existingIndexName);
+          const indexSettings = response.body[existingIndexName].settings.index;
+          const currentNumOfShards = parseInt(indexSettings.number_of_shards);
+          const currentNumOfReplicas = parseInt(
+            indexSettings.number_of_replicas
+          );
+
+          // check if existing latest index settings match default settings
+          if (
+            currentNumOfReplicas ===
+              donorIndexMapping.settings["index.number_of_replicas"] &&
+            currentNumOfShards ===
+              donorIndexMapping.settings["index.number_of_shards"]
+          ) {
+            logger.info(
+              "Existing index settings match default settings, obtaining a new index name from rollcall, clone=true."
+            );
+            newResolvedIndex = await newIndexAndInitializeMapping(
+              rollCallClient,
+              esClient,
+              programId,
+              true
+            );
+          } else {
+            // because existing index settings do not match default, migrate this index
+            logger.info(
+              "Existing index settings do not match default settings, obtaining a new index name from rollcall, clone=false."
+            );
+            newResolvedIndex = await newIndexAndInitializeMapping(
+              rollCallClient,
+              esClient,
+              programId,
+              false
+            );
+
+            logger.info(
+              `Begin reindexing all documents from ${existingIndexName} to ${newResolvedIndex.indexName}`
+            );
+            await esClient.reindex({
+              body: {
+                source: {
+                  index: existingIndexName,
+                },
+                dest: {
+                  index: newResolvedIndex.indexName,
+                },
+              },
+              refresh: true,
+            });
+            logger.info(
+              `Reindexed all documents from ${existingIndexName} to ${newResolvedIndex.indexName}`
+            );
+          }
+        } else {
+          // if no index exists for program, get a new index name
+          logger.info("Obtaining a new index name from rollcall, clone=false.");
+          newResolvedIndex = await newIndexAndInitializeMapping(
+            rollCallClient,
+            esClient,
+            programId,
+            false
+          );
+        }
```

Comment:
this whole chunk seems rather noisy in this file, should we split it into a function that just returns `newResolvedIndex`? `newResolvedIndex` isn't used for anything other than for this purpose so it's kinda leaky...